docs/UpgradingToJest28.md

```diff
@@ -0,0 +1,185 @@
+
+```diff
+- testURL: 'https://jestjs.io'
++ testEnvironmentOptions: {
++   url: 'https://jestjs.io'
++ }
+```
+
+## Fake Timers
+
+Fake timers were refactored to allow passing options to the underlying [`@sinonjs/fake-timers`](https://github.com/sinonjs/fake-timers).
+
+### `fakeTimers`
+
+The `timers` configuration option was renamed to [`fakeTimers`](Configuration.md#faketimers-object) and now takes an object with options:
+
+```diff
+- timers: 'real'
++ fakeTimers: {
++   enableGlobally: false
++ }
+```
+
+```diff
+- timers: 'fake'
++ fakeTimers: {
++   enableGlobally: true
++ }
+```
+
+```diff
+- timers: 'modern'
++ fakeTimers: {
++   enableGlobally: true
++ }
+```
+
+```diff
+- timers: 'legacy'
++ fakeTimers: {
++   enableGlobally: true,
++   legacyFakeTimers: true
++ }
+```
+
+### `jest.useFakeTimers()`
+
+An object with options now should be passed to [`jest.useFakeTimers()`](JestObjectAPI.md#jestusefaketimersfaketimersconfig) as well:
+
+```diff
+- jest.useFakeTimers('modern')
++ jest.useFakeTimers()
+```
+
+```diff
+- jest.useFakeTimers('legacy')
++ jest.useFakeTimers({
++   legacyFakeTimers: true
++ })
+```
+
+If legacy fake timers are enabled in Jest config file, but you would like to disable them in a particular test file:
+
+```diff
+- jest.useFakeTimers('modern')
++ jest.useFakeTimers({
++   legacyFakeTimers: false
++ })
+```
+
+## Test Environment
+
+### Custom Environment
+
+The constructor of [test environment](Configuration.md#testenvironment-string) class now receives an object with Jest's `globalConfig` and `projectConfig` as its first argument. The second argument is now mandatory.
+
+```diff
+  class CustomEnvironment extends NodeEnvironment {
+-   constructor(config) {
+-     super(config);
++   constructor({globalConfig, projectConfig}, context) {
++     super({globalConfig, projectConfig}, context);
++     const config = projectConfig;
+```
+
+### `jsdom`
+
+If you are using JSDOM [test environment](Configuration.md#testenvironment-string), `jest-environment-jsdom` package now must be installed additionally:
+
+```bash npm2yarn
+npm install --save-dev jest-environment-jsdom
+```
+
+## Test Runner
+
+If you are using Jasmine [test runner](Configuration.md#testrunner-string), `jest-jasmine2` package now must be installed additionally:
+
+```bash npm2yarn
+npm install --save-dev jest-jasmine2
+```
+
+## Transformer
+
+`process()` and `processAsync()` methods of a custom [transformer module](CodeTransformation.md) cannot return a string anymore. They must always return an object:
+
+```diff
+  process(sourceText, sourcePath, options) {
+-   return `module.exports = ${JSON.stringify(path.basename(sourcePath))};`;
++   return {
++     code: `module.exports = ${JSON.stringify(path.basename(sourcePath))};`,
++   };
+  }
+```
+
+## TypeScript
+
+:::info
+
+The TypeScript examples from this page will only work as document if you import `jest` from `'@jest/globals'`:
```

Comment:
I assume `@types/jest` will be updated. Something about "until `@types/jest@28` is released"?

Reply:
Yes, updating `@types/jest` would be the best. I just think, if that will be fixed, we will be adding it upgrading guide. Sort-of: "you should use this version of `@types/jest`". So this note will disappear naturally. Perhaps?

Reply:
Reasonable 👍